sdk/java/pulumi/src/main/java/com/pulumi/core/internal/Urn.java

```diff
@@ -190,20 +190,28 @@ public static QualifiedTypeName parse(String qualifiedType) {
                     () -> format("expected qualified type, type part to be not empty, got: '%s'", type)
             );
             final Optional<String> parent;
-            if (qualifiedTypeParts.length == 2) {
-                var pt = qualifiedTypeParts[0];
+
+            // Handle parents. URNs can have nested parents which should be joined with a "$".
+            if (qualifiedTypeParts.length >= 2) {
+                // Remove the last qualifiedTypeParts element as it's the type.
+                String[] parentList = new String[qualifiedTypeParts.length - 1];
+
+                // Get only the parent parts of the URN type.
+                System.arraycopy(qualifiedTypeParts, 0, parentList, 0, parentList.length);
```

Comment:
I think you can shorten this with [Arrays.copyOf](https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html#copyOf-T:A-int-).

```java
var parentList = Arrays.copyOf(qualifiedTypeParts, parentList.length-1)
```

Although seeing that this is just fed into String.join, which accepts an Iterable, you don't need to copy at all. You can create a List view of the array (without copying), and then get a sublist of that:

```java
List<String> parentList = Arrays.asList(qualifiedTypeParts).subList(0, qualifiedTypeParts.length - 1);
var pt = String.join(ParentSeparator, parentList)
```